File: frontend/src/api/dayoff.ts
// 사전 휴무 신청 서버 통신 (스펙 3.1) — 연차와 별개, 기간 단위, 월 20일 한도.
import { apiGet, apiSend } from "./client";

export const MAX_PER_MONTH = 20;

export interface DayOffRequest {
  id: number;
  staff_id: number;
  staff_name: string;
  start_date: string; // "YYYY-MM-DD"
  end_date: string;
  days: number;
  applied_at: string; // 신청일 (쉬는 날짜와는 다른 값)
  status: "requested" | "confirmed" | "rejected";
  note: string | null;
  created_at: string;
}

export interface DayOffRequestCreate {
  staff_id: number;
  start_date: string;
  end_date: string;
  applied_at?: string | null; // 비우면 서버가 오늘 날짜로 저장
  note?: string | null;
}

export const listDayOffRequests = () =>
  apiGet<DayOffRequest[]>("/dayoff-requests");

export const createDayOffRequest = (data: DayOffRequestCreate) =>
  apiSend<DayOffRequest>("POST", "/dayoff-requests", data);

export const updateDayOffRequest = (
  id: number,
  data: Partial<
    Pick<DayOffRequest, "status" | "note" | "start_date" | "end_date" | "applied_at">
  >,
) => apiSend<DayOffRequest>("PATCH", `/dayoff-requests/${id}`, data);

export const deleteDayOffRequest = (id: number) =>
  apiSend<void>("DELETE", `/dayoff-requests/${id}`);

File: backend/app/schemas/dayoff.py
"""사전 휴무 신청 API 데이터 형태 (스펙 3.1) — 기간(시작일~종료일) 단위."""

from datetime import date, datetime

from pydantic import BaseModel

from app.models.enums import LeaveRequestStatus
from app.schemas.leave import _DateRange


class DayOffRequestCreate(_DateRange):
    staff_id: int
    note: str | None = None
    status: LeaveRequestStatus = LeaveRequestStatus.REQUESTED
    applied_at: date | None = None  # 비우면 오늘 날짜로 저장
    reject_reason: str | None = None  # status 가 rejected 일 때 필수


class DayOffRequestUpdate(BaseModel):
    status: LeaveRequestStatus | None = None
    reject_reason: str | None = None  # status 를 rejected 로 바꿀 때 필수
    note: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    applied_at: date | None = None


class DayOffRequestRead(BaseModel):
    id: int
    staff_id: int
    staff_name: str
    start_date: date
    end_date: date
    days: int
    applied_at: date
    status: str
    note: str | None
    reject_reason: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}
